File: myLeetCode/Linked List/19. Remove Nth Node From End of List.py
# 19. Remove Nth Node From End of List
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0, head)
        fast, slow = dummy, dummy
        
        for i in range(n):
            fast = fast.next
        
        while fast.next is not None:
            fast = fast.next
            slow = slow.next
            
        to_delete = slow.next
        slow.next = slow.next.next
        del to_delete
        
        return dummy.next

def main():
    values = [1, 2, 3, 4, 5, 6]
    head = None
    curr = None

    for val in values:
        if head is None:
            head = ListNode(val)
            curr = head
        else:
            curr.next = ListNode(val)
            curr = curr.next

    sol = Solution()
    removed = sol.removeNthFromEnd(head, 2)
    while removed:
        print(removed.val, end=' ')
        removed = removed.next
    print()

if __name__ == "__main__":
    main()

File: myLeetCode/Linked List/19. Remove Nth Node From End of List.cpp
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

#include <iostream>
#include <vector>
struct ListNode {
    int val;
    ListNode *next;

    ListNode(int val) : val(val), next(nullptr) {}
    ListNode(int val, ListNode *next) : val(val), next(next) {}
};

class Solution {
  public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode dummy(0, head);
        ListNode *fast = &dummy, *slow = &dummy;
        for (int i = 1; i <= n; ++i) {
            fast = fast->next;
        }

        while (fast->next != nullptr) {
            fast = fast->next;
            slow = slow->next;
        }

        ListNode *toDelete = slow->next;
        slow->next = slow->next->next;
        delete toDelete;

        return dummy.next;
    }
};

int main() {
    std::vector<int> values = {1, 2, 3, 4, 5, 6};
    ListNode *head = nullptr;
    ListNode *curr = nullptr;

    for (int val : values) {
        if (head == nullptr) {
            head = new ListNode(val);
            curr = head;
        } else {
            curr->next = new ListNode(val);
            curr = curr->next;
        }
    }

    Solution sol;
    ListNode *removed = sol.removeNthFromEnd(head, 2);
    while (removed) {
        std::cout << removed->val << " ";
        removed = removed->next;
    }
    std::cout << std::endl;
    return 0;
}
